Signal/src/ViewControllers/AddToBlockListViewController.m

```diff
@@ -0,0 +1,344 @@
+        instantiateViewControllerWithIdentifier:@"CountryCodeViewController"];
+    countryCodeController.delegate = self;
+    countryCodeController.shouldDismissWithoutSegue = YES;
+    UINavigationController *navigationController =
+        [[UINavigationController alloc] initWithRootViewController:countryCodeController];
+    [self presentViewController:navigationController animated:YES completion:[UIUtil modalCompletionBlock]];
+}
+
+- (void)blockButtonPressed:(id)sender
+{
+    [self tryToBlockPhoneNumber];
+}
+
+- (void)tryToBlockPhoneNumber
+{
+    if (![self hasValidPhoneNumber]) {
+        return;
+    }
+
+    NSString *possiblePhoneNumber = [self.callingCode stringByAppendingString:_phoneNumberTextField.text.digitsOnly];
+    PhoneNumber *parsedPhoneNumber = [PhoneNumber tryParsePhoneNumberFromUserSpecifiedText:possiblePhoneNumber];
+    OWSAssert(parsedPhoneNumber);
+
+    [_blockingManager addBlockedPhoneNumber:[parsedPhoneNumber toE164]];
+
+    UIAlertController *controller = [UIAlertController
+        alertControllerWithTitle:NSLocalizedString(@"BLOCK_LIST_VIEW_BLOCKED_ALERT_TITLE",
+                                     @"The title of the 'phone number blocked' alert in the block view.")
+                         message:[NSString
+                                     stringWithFormat:NSLocalizedString(@"BLOCK_LIST_VIEW_BLOCKED_ALERT_MESSAGE_FORMAT",
+                                                          @"The message format of the 'phone number blocked' alert in "
+                                                          @"the block view. It is populated with the blocked phone "
+                                                          @"number."),
+                                     [parsedPhoneNumber toE164]]
+                  preferredStyle:UIAlertControllerStyleAlert];
+
+    [controller addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"OK", nil)
+                                                   style:UIAlertActionStyleDefault
+                                                 handler:nil]];
+    [self presentViewController:controller animated:YES completion:nil];
+
+    _phoneNumberTextField.text = nil;
+}
+
+- (void)textFieldDidChange:(id)sender
+{
+    [self updateBlockButtonEnabling];
+}
+
+// TODO: We could also do this in registration view.
+- (BOOL)hasValidPhoneNumber
+{
+    if (!self.callingCode) {
+        return NO;
+    }
+    NSString *possiblePhoneNumber = [self.callingCode stringByAppendingString:_phoneNumberTextField.text.digitsOnly];
+    PhoneNumber *parsedPhoneNumber = [PhoneNumber tryParsePhoneNumberFromUserSpecifiedText:possiblePhoneNumber];
+    // It'd be nice to use [PhoneNumber isValid] but it always returns false for some countries
+    // (like afghanistan) and there doesn't seem to be a good way to determine beforehand
+    // which countries it can validate for without forking libPhoneNumber.
+    return parsedPhoneNumber && parsedPhoneNumber.toE164.length > 1;
```

Comment:
This really isn't great.  Users can currently block non-sensical phone numbers like +123.  I'm going to follow up after this PR and see if we can do better.